File: win-prog/ProgWinCS/Tree View and List View/ImageDirectory/ImageDirectory.cs
//---------------------------------------------
// ImageDirectory.cs © 2001 by Charles Petzold
//---------------------------------------------
using Petzold.ProgrammingWindowsWithCSharp;
using System;
using System.Drawing;
using System.Windows.Forms;

class ImageDirectory: Form
{
     PictureBoxPlus    picbox;
     DirectoryTreeView dirtree;
     ImagePanel        imgpanel;
     Splitter          split;
     TreeNode          tnSelect;
     Control           cntlClicked;
     Point             ptPanelAutoScroll;

     public static void Main()
     {
          Application.Run(new ImageDirectory());
     }
     public ImageDirectory()
     {
          Text = "Image Directory";
          BackColor = SystemColors.Window;
          ForeColor = SystemColors.WindowText;

               // Create (invisible) control for displaying large image.

          picbox            = new PictureBoxPlus();
          picbox.Parent     = this;
          picbox.Visible    = false;
          picbox.Dock       = DockStyle.Fill;
          picbox.SizeMode   = PictureBoxSizeMode.StretchImage;
          picbox.NoDistort  = true;
          picbox.MouseDown += new MouseEventHandler(PictureBoxOnMouseDown);

               // Create controls for displaying thumbnails.

          imgpanel        = new ImagePanel();
          imgpanel.Parent = this;
          imgpanel.Dock   = DockStyle.Fill;
          imgpanel.ImageClicked += 
                         new EventHandler(ImagePanelOnImageClicked);

          split           = new Splitter();
          split.Parent    = this;
          split.Dock      = DockStyle.Left;
          split.BackColor = SystemColors.Control;

          dirtree        = new DirectoryTreeView();
          dirtree.Parent = this;
          dirtree.Dock   = DockStyle.Left;
          dirtree.AfterSelect += 
                    new TreeViewEventHandler(DirectoryTreeViewOnAfterSelect);

               // Create menu with one item (Refresh).

          Menu = new MainMenu();
          Menu.MenuItems.Add("&View");
          MenuItem mi = new MenuItem("&Refresh", 
                              new EventHandler(MenuOnRefresh), Shortcut.F5);
          Menu.MenuItems[0].MenuItems.Add(mi);
     }
     void DirectoryTreeViewOnAfterSelect(object obj, TreeViewEventArgs tvea)
     {
          tnSelect = tvea.Node;
          imgpanel.ShowImages(tnSelect.FullPath);
     }
     void MenuOnRefresh(object obj, EventArgs ea)
     {
          dirtree.RefreshTree();
     }
     void ImagePanelOnImageClicked(object obj, EventArgs ea)
     {
               // Get clicked control and image.
          
          cntlClicked = imgpanel.ClickedControl;
          picbox.Image = imgpanel.ClickedImage;

               // Save auto-scroll position.

          ptPanelAutoScroll = imgpanel.AutoScrollPosition;
          ptPanelAutoScroll.X *= -1;
          ptPanelAutoScroll.Y *= -1;

               // Hide and disable the normal controls.

          imgpanel.Visible = false;
          imgpanel.Enabled = false;
          imgpanel.AutoScrollPosition = Point.Empty;

          split.Visible = false;
          split.Enabled = false;

          dirtree.Visible = false;
          dirtree.Enabled = false;

               // Make the picture box visible.

          picbox.Visible = true;
     }
          // Event handlers and method involved with restoring controls

     void PictureBoxOnMouseDown(object obj, MouseEventArgs mea)
     {
          RestoreControls();
     }
     protected override void OnKeyDown(KeyEventArgs kea)
     {
          if (kea.KeyCode == Keys.Escape)
               RestoreControls();
     }
     void RestoreControls()
     {
          picbox.Visible = false;

          dirtree.Visible = true;
          dirtree.Enabled = true;

          split.Enabled = true;
          split.Visible = true;

          imgpanel.AutoScrollPosition = ptPanelAutoScroll;
          imgpanel.Visible = true;
          imgpanel.Enabled = true;
          cntlClicked.Focus();
     }
}

File: win-prog/ProgWinVB/Tree View and List View/ImageDirectory/ImageDirectory.vb
'-----------------------------------------------
' ImageDirectory.vb (c) 2002 by Charles Petzold
'-----------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class ImageDirectory
    Inherits Form

    Private picbox As PictureBoxPlus
    Private dirtree As DirectoryTreeView
    Private imgpnl As ImagePanel
    Private split As Splitter
    Private tnSelect As TreeNode
    Private ctrlClicked As Control
    Private ptPanelAutoScroll As Point

    Shared Sub Main()
        Application.Run(New ImageDirectory())
    End Sub

    Sub New()
        Text = "Image Directory"
        BackColor = SystemColors.Window
        ForeColor = SystemColors.WindowText

        ' Create (invisible) control for displaying large image.
        picbox = New PictureBoxPlus()
        picbox.Parent = Me
        picbox.Visible = False
        picbox.Dock = DockStyle.Fill
        picbox.SizeMode = PictureBoxSizeMode.StretchImage
        picbox.NoDistort = True
        AddHandler picbox.MouseDown, AddressOf PictureBoxOnMouseDown

        ' Create controls for displaying thumbnails.
        imgpnl = New ImagePanel()
        imgpnl.Parent = Me
        imgpnl.Dock = DockStyle.Fill
        AddHandler imgpnl.ImageClicked, AddressOf ImagePanelOnImageClicked

        split = New Splitter()
        split.Parent = Me
        split.Dock = DockStyle.Left
        split.BackColor = SystemColors.Control

        dirtree = New DirectoryTreeView()
        dirtree.Parent = Me
        dirtree.Dock = DockStyle.Left
        AddHandler dirtree.AfterSelect, _
                                    AddressOf DirectoryTreeViewOnAfterSelect

        ' Create menu with one item (Refresh).
        Menu = New MainMenu()
        Menu.MenuItems.Add("&View")
        Dim mi As New MenuItem("&Refresh", _
                               AddressOf MenuOnRefresh, Shortcut.F5)
        Menu.MenuItems(0).MenuItems.Add(mi)
    End Sub

    Private Sub DirectoryTreeViewOnAfterSelect(ByVal obj As Object, _
                                        ByVal tvea As TreeViewEventArgs)
        tnSelect = tvea.Node
        imgpnl.ShowImages(tnSelect.FullPath)
    End Sub

    Private Sub MenuOnRefresh(ByVal obj As Object, ByVal ea As EventArgs)
        dirtree.RefreshTree()
    End Sub

    Private Sub ImagePanelOnImageClicked(ByVal obj As Object, _
                                         ByVal ea As EventArgs)
        ' Get clicked control and image.
        ctrlClicked = imgpnl.ClickedControl
        picbox.Image = imgpnl.ClickedImage

        ' Save auto-scroll position.
        ptPanelAutoScroll = imgpnl.AutoScrollPosition
        ptPanelAutoScroll.X *= -1
        ptPanelAutoScroll.Y *= -1

        ' Hide and disable the normal controls.
        imgpnl.Visible = False
        imgpnl.Enabled = False
        imgpnl.AutoScrollPosition = Point.Empty
        split.Visible = False
        split.Enabled = False
        dirtree.Visible = False
        dirtree.Enabled = False

        ' Make the picture box visible.
        picbox.Visible = True
    End Sub

    ' Event handlers and method involved with restoring controls
    Private Sub PictureBoxOnMouseDown(ByVal obj As Object, _
                                      ByVal mea As MouseEventArgs)
        RestoreControls()
    End Sub

    Protected Overrides Sub OnKeyDown(ByVal kea As KeyEventArgs)
        If kea.KeyCode = Keys.Escape Then
            RestoreControls()
        End If
    End Sub

    Private Sub RestoreControls()
        picbox.Visible = False
        dirtree.Visible = True
        dirtree.Enabled = True
        split.Enabled = True
        split.Visible = True
        imgpnl.AutoScrollPosition = ptPanelAutoScroll
        imgpnl.Visible = True
        imgpnl.Enabled = True
        ctrlClicked.Focus()
    End Sub
End Class
